system/jlib/jevent.cpp

```diff
@@ -1643,65 +1617,61 @@ class CDumpCSVEventVisitor : public CDumpStreamEventVisitor
 
         state = State::Header;
         encodeCSVColumn(markup, "EventName");
-        markup.append(",");
-        encodeCSVColumn(markup, "EventId");
         for(EventAttr a = EventAttr(byte(EvAttrNone) + 1); a < EvAttrMax; a = EventAttr(byte(a) + 1))
         {
-            reverseLookup[queryEventAttributeName(a)] = a;
             markup.append(",");
             encodeCSVColumn(markup, queryEventAttributeName(a));
         }
         dump(true);
         return true;
     }
+
     virtual Continuation visitEvent(EventType id) override
     {
         if (inHeader())
             state = State::Events;
         encodeCSVColumn(markup, queryEventName(id));
-        markup.append(',');
-        StringBuffer tmp;
-        encodeCSVColumn(markup, tmp.append(uint16_t(id)));
-        eventAttrs = headerAttrs;
         return visitContinue;
     }
+
     virtual bool departEvent() override
     {
         for(EventAttr a = EventAttr(byte(EvAttrNone) + 1); a < EvAttrMax; a = EventAttr(byte(a) + 1))
         {
             markup.append(",");
-            Attributes::const_iterator it = eventAttrs.find(a);
-            if (it != eventAttrs.end())
-                encodeCSVColumn(markup, it->second.str());
+            if (row[a].get())
+                encodeCSVColumn(markup, row[a].get());
         }
         dump(true);
-        eventAttrs.clear();
+        for (EventAttr a = EventAttr(byte(EvAttrNone) + 1); a < EvAttrMax; a = EventAttr(byte(a) + 1))
+        {
+            if (row[a].get() && !headerAttrs.count(a))
```

Comment:
I am not clear what the headerAttrs test is there for.

Reply:
Attribute values received before any event (i.e., in the header) must not be cleared after dumping an event row because they can't be replaced. `headerAttr` is updated in  recordAttribute for each attribute in the header, and this check ensures those values persist to the next row.